byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs

```diff
@@ -29,6 +34,44 @@ moneyIsConstant = withTests 200 . property $ do
   (st0, st) <- firstAndLastState <$> forAll (trace @UTXOW 500)
   reserves st0 + balance (utxo st0) === reserves st + balance (utxo st)
 
+-- | Check that a UTxO has no common inputs with a transaction
+noCommonInputs :: UTxO -> Tx -> Bool
+noCommonInputs utxo0 tx = dom (txouts tx) ∩ dom utxo0 == empty
+
+-- | Check that there is no double spending
+noDoubleSpending :: Property
+noDoubleSpending = property $ do
+  t <- forAll (trace @UTXO 500)
+  let
+    utxo0 = fst . firstAndLastState $ t :: State UTXO
+    txs = traceSignals OldestFirst t :: [Signal UTXO]
+  when (all (noCommonInputs $ utxo utxo0) txs) $ do
```

Comment:
Maybe we can replace `noCommonInputs` by its definition, since it makes the predicate look more similar to the specs. Also we do not need the last state it seems. So what about:

```haskell
  let
    UTxOState {utxo = utxo0} = _traceInitState t
    txs = traceSignals OldestFirst t
  when (all (\ti -> dom (txouts ti) ∩ dom utxo0 == empty) txs) 
```

Reply:
I moved `noCommonInputs` to the top level because both the no double spending property and the utxo is outputs minus inputs property use it. With that in mind, do you agree to leave it top-level instead of inlining it in both properties?

Reply:
I would still use the definition, since it looks more similar to the specs. But I'm fine with having `noCommonInputs` if you think that's better.

Just remember to use `_traceInitState`. And I'd remove the type annotations, but again, whatever you prefer.

Reply:
Done! All incorporated (the definition is inlined now).